File: src/medium/BuildTree.go
package main

//输入某二叉树的前序遍历和中序遍历的结果，请构建该二叉树并返回其根节点。
//假设输入的前序遍历和中序遍历的结果中都不含重复的数字。

func main() {

}
func buildTree(preorder []int, inorder []int) *TreeNode {
	ret := &TreeNode{}
	if len(preorder) > 0 {
		bDfs(preorder, inorder, ret)
	} else {
		return nil
	}
	return ret
}
func bDfs(preorder []int, inorder []int, t *TreeNode) {
	x := 0
	//赋值给当前节点
	t.Val = preorder[0]
	if len(preorder) == 1 { //当只剩下一个节点时，赋值后直接返回
		return
	}
	for ; x < len(inorder); x++ {
		//查找左树，中点，右数
		if t.Val == inorder[x] {
			//中点，小于这个点的都是左树，大于这个点的都是右树
			break
		}
	}
	if x != 0 {
		//如果中点在第一个节点，说明没有左子树，就不继续构建
		t.Left = &TreeNode{}
		bDfs(preorder[1:x+1], inorder[:x], t.Left)
	}
	if x != len(inorder)-1 {
		//如果中点在最后一个节点，说明没有右子树，就不继续构建
		t.Right = &TreeNode{}
		bDfs(preorder[x+1:], inorder[x+1:], t.Right)
	}
}

File: src/medium/BuildTree.java
package medium;

import java.util.Arrays;

/**
 * @author siianchan
 * @since 2024/2/21 13:41
 */
public class BuildTree {

    /**
     * 给定两个整数数组 inorder 和 postorder ，其中 inorder 是二叉树的中序遍历
     * ， postorder 是同一棵树的后序遍历，请你构造并返回这颗 二叉树 。
     * 输入：inorder = [9,3,15,20,7], postorder = [9,15,7,20,3]
     * 输出：[3,9,20,null,null,15,7]
     */
    public static void main(String[] args) {
        buildTree(new int[]{9, 3, 15, 20, 7}, new int[]{9, 15, 7, 20, 3});
    }

    public static TreeNode buildTree(int[] inorder, int[] postorder) {
        if (postorder == null || postorder.length < 1) {
            //节点为空就返回null
            return null;
        }
        int inorderRootIndex = 0;
        //最右边一个是根节点
        int root = postorder[postorder.length - 1];
        //构建当前节点（根节点）
        TreeNode treeNode = new TreeNode(root);
        for (int x = 0; x < inorder.length; x++) {
            //找到根节点在中序遍历中的位置，他左边都是左子树，右边都是右子树
            if (inorder[x] == root) {
                inorderRootIndex = x;
                break;
            }
        }
        //把左右子树分开来，然后递归处理
        int[] leftInorder = Arrays.copyOfRange(inorder, 0, inorderRootIndex);
        int[] rightInorder = Arrays.copyOfRange(inorder, inorderRootIndex + 1, inorder.length);
        int[] leftPostorder = Arrays.copyOfRange(postorder, 0, leftInorder.length);
        int[] rightPostorder = Arrays.copyOfRange(postorder, leftInorder.length, leftInorder.length + rightInorder.length);
        //递归构建左节点
        treeNode.left = buildTree(leftInorder, leftPostorder);
        //递归构建右节点
        treeNode.right = buildTree(rightInorder, rightPostorder);
        return treeNode;
    }

}
